first pass at @link documentation (#1693)
apollographql/federation: docs/source/federation-spec.md

> Note: Repeated directives (in this case, `@key`, used multiple times) require support by the underlying GraphQL implementation.

### `@link`

```graphql
directive @link(
  url: String, 
  as: String, 
  for: link__Purpose, 
  import: [link__Import]
) repeatable on SCHEMA
```

> The `@link` directive links definitions within the document to external schemas. External schemas are identified by their `url`, which optionally ends with a name and version in the form `name`/v`major`.`minor`.

Reviewer: I think the deploy preview made it seem a little unclear what the URL should be, I'd recommend splitting this out into two different sentences, like so:

```
External schemas are identified by their `url`, which optionally ends with a name and version with the following format:

`{NAME}`/v`{MAJOR}`.`{MINOR}`
```

The presence of a `@link` directive makes a document a [core schema](https://specs.apollo.dev/#def-core-schema).

The `for` argument describes the purpose of a `@link`. Currently accepted values are `SECURITY` or `EXECUTION`. Core schema-aware servers such as Apollo Router and Gateway will refuse to operate on schemas which contain `@link`s to unsupported specs which are `for: SECURITY` or `for: EXECUTION`.

By default, `@link`ed definitions will be namespaced, i.e. `@federation__requires`. The `as` argument lets you pick the name for this namespace:
```graphql

> extend schema @link(url: "https://specs.apollo.dev/federation/v2.0". as: "fed2")

Author: should be a comma, good catch.

```

This type extension in the Reviews service extends the `User` type from the Users service. It extends it for the purpose of adding a new field called `reviews`, which returns a list of `Review`s.

> 

Author: Was thinking along the same lines. I'll move it up.

### `@link`

```graphql
scalar link__Import

enum link__Purpose {
  """
  \`SECURITY\` features provide metadata necessary to securely resolve fields.
  """
  SECURITY

  """
  \`EXECUTION\` features provide metadata necessary for operation execution.
  """
  EXECUTION
}

directive @link(
  url: String, 
  as: String, 
  for: link__Purpose, 
  import: [link__Import]
) repeatable on SCHEMA
```

> The `@link` directive provides a mechanism to link to external schemas, specifying the URL, version, and directives. The presence of a `@link` directive also indicates that the schema is a core schema. 

Reviewer 2: ```suggestion
The `@link` directive links definitions within the document to external schemas. External schemas are identified by their `url`, which optionally ends with a name and version in the form `name`/v`major`.`minor`.

The presence of a `@link` directive makes a document a [core schema](https://specs.apollo.dev/#def-core-schema).

The `for` argument describes the purpose of a `@link`. Currently accepted values are `SECURITY` or `EXECUTION`. Core schema-aware servers such as Apollo Router and Gateway will refuse to operate on schemas which contain `@link`s to unsupported specs which are `for: SECURITY` or `for: EXECUTION`.

By default, `@link`ed definitions will be namespaced, i.e. `@federation__requires`. The `import` argument lets you import external definitions into your namespace, so they lose the prefix:
```

Reviewer 3: @queerviolet Should that last bit be
> The `as` argument lets you import

I might also word golf that a little
> By default, `@link`ed definitions are namespaced with the name of the schema, e.g. `@federation__requires`. The `as` argument lets you import definitions without the prefix.

Reviewer 3: @queerviolet Should that last bit be
> The `as` argument lets you import

I might also word golf that a little if we're open to it:

> By default, `@link`ed definitions are namespaced with the name of the schema, e.g. `@federation__requires`. The `as` argument lets you import definitions without the prefix.

```graphql
    schema
      @link(url: "https://specs.apollo.dev/link/v1.0")

>       @link(url: "https://specs.apollo.dev/federation/v2.0", import: ["@key", "@requires", "@provides", "@external", { name: "@tag", as "@mytag" }, "@extends", "@shareable", "@inaccessible", "@override"])

Reviewer 2: ```suggestion
      @link(url: "https://specs.apollo.dev/federation/v2.0", import: ["@key", "@requires", "@provides", "@external", { name: "@tag", as: "@mytag" }, "@extends", "@shareable", "@inaccessible", "@override"])
```

```
 

> When present, the `import` argument will assign non-prefixed (local) names to elements from a foreign schema. In the example above, `@tag` is renamed to `@mytag` while the other directive names will be assigned as they are specified in the federation schema. The ability to rename directives may be necessary if there is a naming collision in two different foreign schemas.

Reviewer 4: Do you think talking about redefining directives in here would be useful or purposely leaving it out is the right way to go?  I do not have a strong opinion, but just musing.

Reviewer 2: does renaming directives work? we should maybe not talk about it unless it works e2e

Reviewer 2: ```suggestion
In the example above, we import various directives from `federation/v2.0` into our namespace. We also rename one of them, bringing in federation's `@tag` as `@mytag` to distinguish it from a different `@tag` directive already in the schema.
```

### `@link`

```graphql
scalar link__Import

enum link__Purpose {
  """
  \`SECURITY\` features provide metadata necessary to securely resolve fields.
  """
  SECURITY

  """
  \`EXECUTION\` features provide metadata necessary for operation execution.
  """
  EXECUTION
}

directive @link(
  url: String, 
  as: String, 
  for: link__Purpose, 
  import: [link__Import]
) repeatable on SCHEMA
```

The `@link` directive links definitions within the document to external schemas. External schemas are identified by their `url`, which optionally ends with a name and version in the form `name`/v`major`.`minor`.

The presence of a `@link` directive makes a document a [core schema](https://specs.apollo.dev/#def-core-schema).

The `for` argument describes the purpose of a `@link`. Currently accepted values are `SECURITY` or `EXECUTION`. Core schema-aware servers such as Apollo Router and Gateway will refuse to operate on schemas which contain `@link`s to unsupported specs which are `for: SECURITY` or `for: EXECUTION`.

> By default, `@link`ed definitions will be namespaced, i.e. `@federation__requires`. The `import` argument lets you import external definitions into your namespace, so they lose the prefix:

Reviewer 2: ```suggestion
By default, `@link`ed definitions will be namespaced, i.e. `@federation__requires`. The `as` argument lets you pick the name for this namespace:

```graphql
extend schema @link(url: "https://specs.apollo.dev/federation/v2.0". as: "fed2")

type User {
  metrics: Metrics @external
  favorites: UserFavorites @fed2__requires(fields: "metrics")
}
```

The `import` argument lets you import external definitions into your namespace, so they lose the prefix:
```
